data_structures/src/chain.rs

```diff
@@ -164,6 +165,202 @@ pub type SHA256 = [u8; 32];
+    ///
+    /// The `key` may be any borrowed form of the hash, but `Hash` and
+    /// `Eq` on the borrowed form must match those for the key type.
+    ///
+    /// # Examples:
+    /// ```
+    /// # use witnet_data_structures::chain::{TransactionsPool, Transaction, Hash};
+    /// let mut pool = TransactionsPool::new();
+    /// let hash = Hash::SHA256([0 as u8; 32]);
+    ///
+    /// assert!(!pool.contains(&hash));
+    ///
+    /// pool.insert(hash, Transaction);
+    ///
+    /// assert!(pool.contains(&hash));
+    /// ```
+    pub fn contains(&self, key: &Hash) -> bool {
+        self.transactions.contains_key(key)
+    }
+
+    /// Insert a transaction identified by `key` into the pool.
+    ///
+    /// # Examples:
+    ///
+    /// ```
+    /// # use witnet_data_structures::chain::{TransactionsPool, Transaction, Hash};
+    /// let mut pool = TransactionsPool::new();
+    ///
+    /// pool.insert(Hash::SHA256([0 as u8; 32]), Transaction);
+    ///
+    /// assert!(!pool.is_empty());
+    /// ```
+    pub fn insert(&mut self, key: Hash, transaction: Transaction) {
+        let weight = 0; // TODO: weight = transaction-fee / transaction-weight
+        self.transactions.insert(key, (weight, transaction));
+        self.sorted_index.insert((weight, key));
+    }
+
+    /// An iterator visiting all the transactions in the pool in
+    /// descending-fee order, that is, transactions with bigger fees
+    /// come first.
+    ///
+    /// Examples:
+    ///
+    /// ```
+    /// # use witnet_data_structures::chain::{TransactionsPool, Transaction, Hash};
+    /// let mut pool = TransactionsPool::new();
+    ///
+    /// pool.insert(Hash::SHA256([0 as u8; 32]), Transaction);
+    /// pool.insert(Hash::SHA256([0 as u8; 32]), Transaction);
+    ///
+    /// let mut iter = pool.iter();
+    /// let tx1 = iter.next();
+    /// let tx2 = iter.next();
+    ///
+    /// // TODO: assert!(tx1.weight() >= tx2.weight());
+    /// ```
+    pub fn iter(&self) -> impl Iterator<Item = &Transaction> {
+        self.sorted_index
+            .iter()
+            .rev()
+            .filter_map(move |(_, h)| self.transactions.get(h).map(|(_, t)| t))
+    }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all transactions such that
+    /// `f(&Hash,&mut Transaction)` returns `false`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use witnet_data_structures::chain::{TransactionsPool, Transaction, Hash};
+    ///
+    /// let mut pool = TransactionsPool::new();
+    /// pool.insert(Hash::SHA256([0 as u8; 32]), Transaction);
+    /// pool.insert(Hash::SHA256([1 as u8; 32]), Transaction);
+    /// assert_eq!(pool.len(), 2);
+    /// pool.retain(|h, _| match h { Hash::SHA256(n) => n[0]== 0 });
+    /// assert_eq!(pool.len(), 1);
+    /// ```
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&Hash, &mut Transaction) -> bool,
```

Comment:
The retain method is more useful if we allow to mutate the values, but that conflicts with our cache idea (because we would need to check for every item that did change and recalculate the cache). Also, the transactions are inherently immutable so I think a good compromise would be:

```suggestion
        F: FnMut(&Hash, &Transaction) -> bool,
```

Reply:
It should be `FnMut`, not `Fn`, like

https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain